### PR TITLE
[GSoC] Add ZSTD compression algorithm to ROOT I/O

### PR DESCRIPTION
**Pull request by fylux**

### CODE REVIEW

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@phsft-bot  build!

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by fylux** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
> @phsft-bot build!

sorry you don't have rights to trigger ROOT builds..

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by oshadura** [comment]
@bbockelm we can start to review PR.. I cant reproduce failures on ROOT-performance-centos7-multicore/default both as separate build on node and or even in build directory, I suspect it could be a problem of "very" parallel  builds (-j56). I will test and let you know..

**Comment by bellenot** [comment]
@fylux please fork also roottest in order to enable building on Windows

**Comment by fylux** [comment]
> @fylux please fork also roottest in order to enable building on Windows

Done! https://github.com/fylux/roottest/tree/zstd-dev2

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@bbockelm failures on windows are not connected..
07:05:13 C:\build\workspace\root-pullrequests-build\root\test\testTBB.cxx(1): fatal error C1083: Cannot open include file: 'tbb/parallel_for.h': No such file or directory [C:\build\workspace\root-pullrequests-build\build\test\testTBB.vcxproj]

File: cmake/modules/SearchInstalledSoftware.cmake [hunk 1]
```diff
@@ -219,6 +219,45 @@ if(builtin_xxhash)
...
+endif()
+
+if (builtin_zstd)
+  set(zstd_version v1.4.0)
+  set(ZSTD_TARGET ZSTD)
+  message(STATUS "Building ZSTD version ${zstd_version} included in ROOT itself")
+  set(ZSTD_URL https://github.com/facebook/zstd/archive/v1.4.0.tar.gz)
+  set(ZSTD_SHA256 63be339137d2b683c6d19a9e34f4fb684790e864fee13c7dd40e197a64c705c1)
+  set(ZSTD_LIBRARIES ${CMAKE_BINARY_DIR}/lib/${CMAKE_STATIC_LIBRARY_PREFIX}zstd${CMAKE_STATIC_LIBRARY_SUFFIX})
+  set(ZSTD_C_FLAGS "-fPIC -Wno-unused-variable -O3")
```
Comment (bellenot) [inline cmake/modules/SearchInstalledSoftware.cmake]: Those flags are incompatible with the Microsoft compiler

Comment (fylux) [reply cmake/modules/SearchInstalledSoftware.cmake]: Fixed

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@fylux  ehat about clang-formatting? Can you add please as a separate commit? Logs are here https://travis-ci.org/root-project/root/jobs/553883729

**Comment by oshadura** [comment]
@fylux Jenkins build is successful! Congrats! After clang-formatting, PR will be ready for review.

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
LGTM! (after showing/presenting performance results)

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@fylux can you squash two last commits together please?

File: cmake/modules/FindZSTD.cmake [hunk 2]
```diff
@@ -0,0 +1,79 @@
+
+#
+# FindZSTD.cmake
```
[the remark this answers is not in the document]
Comment (oshadura) [reply cmake/modules/FindZSTD.cmake]: Ok I will add it on the top, when Alfonso @fylux will finish with small fixes..

File: core/zip/src/RZip.cxx [hunk 3]
```diff
@@ -314,102 +327,112 @@ int R__unzip_header(int *srcsize, uch *src, int *tgtsize)
...
+   }
+
+   if (ibufcnt + HDRSIZE != *srcsize) {
+      fprintf(stderr, "R__unzip: discrepancy in source length\n");
+      return;
+   }
+
+   /*   D E C O M P R E S S   D A T A  */
+
+   /*std::cout << "tgtsize: " << *tgtsize << std::endl;
```
[the remark this answers is not in the document]
Comment (fylux) [reply core/zip/src/RZip.cxx]: Fixed

File: core/zstd/src/ZipZSTD.cxx [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+#include <zstd.h>
+#include <memory>
+
+#include <iostream>
+
+static const int kHeaderSize = 9;
+
+void R__zipZSTD(int cxlevel, int *srcsize, char *src, int *tgtsize, char *tgt, int *irep)
+{
+   using Ctx_ptr = std::unique_ptr<ZSTD_CCtx, decltype(&ZSTD_freeCCtx)>;
```
[the remark this answers is not in the document]
Comment (fylux) [reply core/zstd/src/ZipZSTD.cxx]: Fixed

File: core/zip/inc/Compression.h [hunk 5]
```diff
@@ -87,6 +94,8 @@ struct RCompressionSetting {
          kOldCompressionAlgo,
          /// Use LZ4 compression
          kLZ4,
+         /// Use ZSTD compression
+         kZSTD, ///
```
[the remark this answers is not in the document]
Comment (fylux) [reply core/zip/inc/Compression.h]: Fixed

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
Code was merged in other PR

**Pull request closed without merging**